scene/gui/color_picker.cpp

```diff
@@ -527,45 +597,109 @@ inline int ColorPicker::_get_preset_size() {
+void ColorPicker::_set_mode_popup_value(ColorModeType p_mode) {
+	ERR_FAIL_INDEX(p_mode, MODE_MAX + 1);
 
-	if (!is_inside_tree()) {
+	if (p_mode == MODE_MAX) {
+		set_colorize_sliders(!colorize_sliders);
+	} else {
+		set_color_mode(p_mode);
+	}
+}
+
+int ColorPicker::_get_preset_idx_at_point(const Point2 &p_point) {
+	int hover_now = -1;
+	for (int i = 0; i < preset_container->get_child_count(); i++) {
+		Rect2 rect = _get_preset_rect(i);
+		if (rect.has_point(p_point)) {
+			hover_now = i;
+		}
+	}
+
+	return hover_now;
+}
+
+Variant ColorPicker::_get_drag_data_fw(const Point2 &p_point, Control *p_from_control) {
+	ColorPresetButton *dragged_preset_button = Object::cast_to<ColorPresetButton>(p_from_control);
+
+	if (!dragged_preset_button) {
+		return Variant();
+	}
+
+	int dragged_index = dragged_preset_button->get_index();
+
+	if (dragged_index < 0) {
+		return Variant();
+	}
+
+	ColorPresetButton *drag_preview = memnew(ColorPresetButton(dragged_preset_button->get_preset_color()));
+	drag_preview->set_custom_minimum_size(Size2(_get_preset_size(), _get_preset_size()));
+	set_drag_preview(drag_preview);
+
+	Dictionary drag_data;
+	drag_data["type"] = "color_preset";
+	drag_data["color_preset"] = dragged_index;
+
+	return drag_data;
+}
+
+bool ColorPicker::_can_drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from_control) const {
+	Dictionary d = p_data;
+	if (!d.has("type") || String(d["type"]) != "color_preset") {
+		return false;
+	}
+	return true;
+}
+
+void ColorPicker::_drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from_control) {
+	Dictionary d = p_data;
+	if (!d.has("type")) {
 		return;
 	}
 
-	_update_controls();
-	_update_color();
+	if (String(d["type"]) == "color_preset") {
+		int preset_from_id = d["color_preset"];
+		int hover_now = _get_preset_idx_at_point(p_from_control->get_position() + p_point);
+
+		if (preset_from_id == hover_now || hover_now == -1) {
+			return;
+		}
+		preset_container->move_child(preset_container->get_child(preset_from_id), hover_now);
+	}
 }
 
 void ColorPicker::add_preset(const Color &p_color) {
-	if (presets.find(p_color)) {
+	if (selected_preset) {
 		presets.move_to_back(presets.find(p_color));
-
-		// Find button to move to the end.
-		for (int i = 1; i < preset_container->get_child_count(); i++) {
-			ColorPresetButton *current_btn = Object::cast_to<ColorPresetButton>(preset_container->get_child(i));
-			if (current_btn && p_color == current_btn->get_preset_color()) {
-				preset_container->move_child(current_btn, preset_container->get_child_count() - 1);
-				break;
-			}
-		}
+		preset_container->move_child(selected_preset, preset_container->get_child_count() - 1);
 	} else {
 		presets.push_back(p_color);
 		preset_cache.push_back(p_color);
 
 		_add_preset_button(_get_preset_size(), p_color);
+		ColorPresetButton *last_preset_button = static_cast<ColorPresetButton *>(preset_container->get_child(preset_container->get_child_count() - 1));
```

Comment:
```suggestion
		ColorPresetButton *last_preset_button = Object::cast_to<ColorPresetButton>(preset_container->get_child(preset_container->get_child_count() - 1));
```
Why not like this?

Reply:
> Why not like this?

What's the difference?

Reply:
Functionally none, the latter is just preferred  in the code base.